File: decompiled/sources/com/hjq/shape/other/LinearGradientFontSpan.java
package com.hjq.shape.other;

import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Shader;
import android.text.SpannableStringBuilder;
import android.text.style.ReplacementSpan;

public class LinearGradientFontSpan extends ReplacementSpan {
    public static final int GRADIENT_ORIENTATION_HORIZONTAL = 0;
    public static final int GRADIENT_ORIENTATION_VERTICAL = 1;
    private float mMeasureTextWidth;
    private int[] mTextGradientColors;
    private int mTextGradientOrientation;
    private float[] mTextGradientPositions;

    public static SpannableStringBuilder buildLinearGradientSpannable(CharSequence charSequence, int[] iArr, float[] fArr, int i) {
        SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder(charSequence);
        LinearGradientFontSpan linearGradientFontSpan = new LinearGradientFontSpan();
        linearGradientFontSpan.setTextGradientColor(iArr);
        linearGradientFontSpan.setTextGradientOrientation(i);
        linearGradientFontSpan.setTextGradientPositions(fArr);
        spannableStringBuilder.setSpan(linearGradientFontSpan, 0, spannableStringBuilder.length(), 33);
        return spannableStringBuilder;
    }

    public int getSize(Paint paint, CharSequence charSequence, int i, int i2, Paint.FontMetricsInt fontMetricsInt) {
        this.mMeasureTextWidth = paint.measureText(charSequence, i, i2);
        Paint.FontMetricsInt fontMetricsInt2 = paint.getFontMetricsInt();
        if (fontMetricsInt != null) {
            fontMetricsInt.top = fontMetricsInt2.top;
            fontMetricsInt.ascent = fontMetricsInt2.ascent;
            fontMetricsInt.descent = fontMetricsInt2.descent;
            fontMetricsInt.bottom = fontMetricsInt2.bottom;
        }
        return (int) this.mMeasureTextWidth;
    }

    public void draw(Canvas canvas, CharSequence charSequence, int i, int i2, float f, int i3, int i4, int i5, Paint paint) {
        LinearGradient linearGradient;
        Paint paint2 = paint;
        if (this.mTextGradientOrientation == 1) {
            linearGradient = new LinearGradient(0.0f, 0.0f, 0.0f, paint.descent() - paint.ascent(), this.mTextGradientColors, this.mTextGradientPositions, Shader.TileMode.REPEAT);
        } else {
            float f2 = f + this.mMeasureTextWidth;
            int[] iArr = this.mTextGradientColors;
            linearGradient = new LinearGradient(f, 0.0f, f2, 0.0f, iArr, this.mTextGradientPositions, Shader.TileMode.REPEAT);
        }
        paint2.setShader(linearGradient);
        int alpha = paint.getAlpha();
        if (alpha != 255) {
            paint2.setAlpha(255);
        }
        canvas.drawText(charSequence, i, i2, f, (float) i4, paint);
        paint2.setAlpha(alpha);
    }

    public LinearGradientFontSpan setTextGradientOrientation(int i) {
        this.mTextGradientOrientation = i;
        return this;
    }

    public LinearGradientFontSpan setTextGradientColor(int[] iArr) {
        this.mTextGradientColors = iArr;
        return this;
    }

    public LinearGradientFontSpan setTextGradientPositions(float[] fArr) {
        this.mTextGradientPositions = fArr;
        return this;
    }
}

File: decompiled/sources/com/hjq/shape/other/LinearGradientFontSpan.kt
package com.hjq.shape.other

import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Shader
import android.text.SpannableStringBuilder
import android.text.style.ReplacementSpan

class LinearGradientFontSpan : ReplacementSpan() {
    private var mMeasureTextWidth = 0f
    private var mTextGradientColors: IntArray? = null
    private var mTextGradientOrientation = 0
    private var mTextGradientPositions: FloatArray? = null

    override fun getSize(
        paint: Paint,
        text: CharSequence,
        start: Int,
        end: Int,
        fm: Paint.FontMetricsInt?
    ): Int {
        mMeasureTextWidth = paint.measureText(text, start, end)
        val fontMetrics = paint.fontMetricsInt
        if (fm != null) {
            fm.top = fontMetrics.top
            fm.ascent = fontMetrics.ascent
            fm.descent = fontMetrics.descent
            fm.bottom = fontMetrics.bottom
        }
        return mMeasureTextWidth.toInt()
    }

    override fun draw(
        canvas: Canvas,
        text: CharSequence,
        start: Int,
        end: Int,
        x: Float,
        top: Int,
        y: Int,
        bottom: Int,
        paint: Paint
    ) {
        val linearGradient =
            if (mTextGradientOrientation == GRADIENT_ORIENTATION_VERTICAL) {
                LinearGradient(
                    0f,
                    0f,
                    0f,
                    paint.descent() - paint.ascent(),
                    mTextGradientColors!!,
                    mTextGradientPositions,
                    Shader.TileMode.REPEAT
                )
            } else {
                LinearGradient(
                    x,
                    0f,
                    x + mMeasureTextWidth,
                    0f,
                    mTextGradientColors!!,
                    mTextGradientPositions,
                    Shader.TileMode.REPEAT
                )
            }
        paint.shader = linearGradient
        val alpha = paint.alpha
        if (alpha != 255) {
            paint.alpha = 255
        }
        canvas.drawText(text, start, end, x, y.toFloat(), paint)
        paint.alpha = alpha
    }

    fun setTextGradientOrientation(orientation: Int): LinearGradientFontSpan {
        mTextGradientOrientation = orientation
        return this
    }

    fun setTextGradientColor(colors: IntArray?): LinearGradientFontSpan {
        mTextGradientColors = colors
        return this
    }

    fun setTextGradientPositions(positions: FloatArray?): LinearGradientFontSpan {
        mTextGradientPositions = positions
        return this
    }

    companion object {
        const val GRADIENT_ORIENTATION_HORIZONTAL = 0
        const val GRADIENT_ORIENTATION_VERTICAL = 1

        @JvmStatic
        fun buildLinearGradientSpannable(
            text: CharSequence,
            colors: IntArray?,
            positions: FloatArray?,
            orientation: Int
        ): SpannableStringBuilder {
            val spannable = SpannableStringBuilder(text)
            val span =
                LinearGradientFontSpan()
                    .setTextGradientColor(colors)
                    .setTextGradientOrientation(orientation)
                    .setTextGradientPositions(positions)
            spannable.setSpan(span, 0, spannable.length, 33)
            return spannable
        }
    }
}
